core/pkg/monitor/openmetrics.go

```diff
@@ -0,0 +1,279 @@
+	for i, filter := range sequence.GetValue() {
+		result[i] = Filter{MetricNameRegex: filter}
+	}
+	return result
+}
+
+// processMappingFilters converts a mapping of metric regex patterns to label regex patterns
+// to a list of Filter objects.
+func processMappingFilters(mapping *spb.MapStringKeyMapStringKeyStringValue) []Filter {
+	if mapping == nil {
+		return nil
+	}
+	result := []Filter{}
+	for metricRegex, labelFilters := range mapping.GetValue() {
+		filter := Filter{MetricNameRegex: metricRegex}
+		for labelName, labelRegex := range labelFilters.GetValue() {
+			filter.LabelFilters = append(filter.LabelFilters, [2]string{labelName, labelRegex})
+		}
+		result = append(result, filter)
+	}
+	return result
+}
+
+// Constants for the default OpenMetrics HTTP client configuration.
+//
+// Metrics are collected on a best-effort basis, but we do allow for some retries.
+const (
+	DefaultOpenMetricsRetryMax     = 3
+	DefaultOpenMetricsRetryWaitMin = 1 * time.Second
+	DefaultOpenMetricsRetryWaitMax = 10 * time.Second
+	DefaultOpenMetricsTimeout      = 5 * time.Second
+)
+
+// OpenMetrics is a monitor that collects metrics from an OpenMetrics endpoint.
+//
+// Supports only GAUGE and COUNTER metrics.
+type OpenMetrics struct {
+	name        string
+	url         string
+	filters     []Filter
+	client      *retryablehttp.Client
+	logger      *observability.CoreLogger
+	labelMap    map[string]map[string]int
+	labelHashes map[string]map[string]string
+}
+
+func NewOpenMetrics(
+	logger *observability.CoreLogger,
+	name string,
+	url string,
+	filters *spb.OpenMetricsFilters,
+	retryClient *retryablehttp.Client,
+) *OpenMetrics {
+	var client *retryablehttp.Client
+
+	if retryClient == nil {
+		client = retryablehttp.NewClient()
+		client.Logger = logger
+		client.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy
+		client.RetryMax = DefaultOpenMetricsRetryMax
+		client.RetryWaitMin = DefaultOpenMetricsRetryWaitMin
+		client.RetryWaitMax = DefaultOpenMetricsRetryWaitMax
+		client.HTTPClient.Timeout = DefaultOpenMetricsTimeout
+		client.Backoff = clients.ExponentialBackoffWithJitter
+	} else {
+		client = retryClient
+	}
+
+	var processedFilters []Filter
+
+	if filters != nil {
+		switch v := filters.GetValue().(type) {
+		case *spb.OpenMetricsFilters_Sequence:
+			processedFilters = processSequenceFilters(v.Sequence)
+		case *spb.OpenMetricsFilters_Mapping:
+			processedFilters = processMappingFilters(v.Mapping)
+		default:
+			logger.Warn("Unknown filter type, using empty filter")
+		}
+	}
+
+	om := &OpenMetrics{
+		name:        name,
+		url:         url,
+		filters:     processedFilters,
+		client:      client,
+		logger:      logger,
+		labelMap:    make(map[string]map[string]int),
+		labelHashes: make(map[string]map[string]string),
+	}
+
+	return om
+}
+
+func (o *OpenMetrics) Name() string { return o.name }
+
+// ShouldCaptureMetric checks if a metric should be captured based on the filters.
+func ShouldCaptureMetric(
```

Comment:
maybe having some caching mechanism would be good here, if i understand correctly this will match every time?